Frontend rewrite using Angular

hrueger opened the pull request:
**Advantages**:
* HTML elements are not created in strings and then added to the DOM, instead Angular handles all of that based on the template and its logic
* Removed the call to the external QR Code Server as it is now generated in the frontend
* No code duplication (e.g. navbar)
* improved credentials form
* Listener Clients in Producer View are sorted based on their Active / Inactive Status
* Complete typings for all Objects
* Improved UI with different colored buttons and icons
* Loading Spinners
* Progressive Web App functionality (e.g. to Home Screen)
* better confirm dialogs
* Chat History is kept until page refresh

**Current Status**:
- [x] Base structure
- [x] Home Screen
- [x] Tally Screen
- [x] Producer Screen
- [x] Settings Screen
- [x] Cloud Functionality
- [x] Icons
- [x] Fix no device states on initial load (producer screen)
- [x] Move sockets into service
- [x] SweetAlert
- [x] Loading indicators
- [x] Separate About Page
- [x] Redesign Home Page
- [x] Ports in use
- [x] Validation
- [x] Authentication
- [x] Chat
- [x] NoSleep
- [x] Favicon
- [x] PWA Stuff
- [x] Build Process
	- [x] CLI
	- [x] Electron
- [x] GitHub Actions

fixes #86

hrueger commented:
@josephdadams I have thought about the release process. If I read the `README.md` correctly, currently it can only be installed by cloning or downloading the GitHub Repository, installing the dependencies and then running `node index.js`.

Here are my ideas (feel free to add some or criticize mine ;-) )

* **Docker Image**: It could be run in the cloud and on every machine which has Docker installed with `docker run tallyarbiter -p 4455:4455`. 
* **NPM Package**: Could be run with `npm i -g tallyarbiter` and then `tallyarbiter`
* **Electron + Installer**: With Electron we could make a full-featured cross-platform desktop app. Users could download the installer and the run TallyArbiter like and other software. Could be great for users with less terminal experience. However, I don't know if this works on a Raspberry Pi with the lite image (so without the GUI)

What do you think? With GitHub Actions it would not be a problem to create a GitHub Release automatically on every git tag. We could go with all the release options to provide maximum flexibility for users.

josephdadams commented:
@hrueger great thoughts! The electron package is definitely something I’ve wanted to do but I’ve been holding off until the codebase was more manageable. I even prototyped it and had it working fine but didn’t want to release it until I had done something to compartmentalize the areas of code.

I’m fine with the other methods as well. 

You’re correct that electron won’t run on Pi’s without desktops, and pi compatibility is important to me. Bitfocus Companion achieves this through a headless script, so perhaps we could model after that.

hrueger commented:
>You’re correct that electron won’t run on Pi’s without desktops, and pi compatibility is important to me. Bitfocus Companion achieves this through a headless script, so perhaps we could model after that.

Could Pi uses just use the `tallyarbiter` npm package instead of the electron version?

hrueger commented:
@josephdadams  Is there a reason why the `navigator.vibrate` method is called like this:
```TypeScript
let successBool = window.navigator.vibrate(100, 30, 100, 30, 100);
```
The TS Autocompletion wants
```TypeScript
let successBool = window.navigator.vibrate([100, 30, 100, 30, 100]);
```

Are there devices which don't want an Array here?

josephdadams commented:
> > You’re correct that electron won’t run on Pi’s without desktops, and pi compatibility is important to me. Bitfocus Companion achieves this through a headless script, so perhaps we could model after that.
> 
> Could Pi uses just use the `tallyarbiter` npm package instead of the electron version?

I believe so.

josephdadams commented:
> @josephdadams Is there a reason why the `navigator.vibrate` method is called like this:
> 
> ```ts
> let successBool = window.navigator.vibrate(100, 30, 100, 30, 100);
> ```
> 
> The TS Autocompletion wants
> 
> ```ts
> let successBool = window.navigator.vibrate([100, 30, 100, 30, 100]);
> ```
> 
> Are there devices which don't want an Array here?

Tbh I probably copied that from somewhere else that way. So whatever works.

hrueger commented:
Is there already some logic in place to identify web clients? If I refresh my Tally Page (e.g. with F5), a new listener appears in the producer / settings view and the old one gets greyed out.

I'd create a unique ID on the client side and save this in localStorage. Then, I could send this to the socket server which could identify based on that id instead of the socket id. What do you think? 

This is probably needed in order to save settings for each client (like in #11).

JTF4 commented:
I think that would require a recode of how all of the clients handle their ID. Is there any way to make it so that clients that support local storage use your method and older versions use the current system?

I for one would be all for a system that can rematch disconnected clients with their last assigned source though.

josephdadams commented:
My thinking had been:
1. Listener connects
2. Server generates id and sends to client
3. Client saves this id and sends it on subsequent requests
4. If the id is not valid for some reason, it is just treated like a new connection and a new id is made. 

We can use this same id on the server side to send preferences like preferred device id, colors, etc. This way the server can relate the id to the type and enabled features

hrueger commented:
Sounds good. As this also requires changes in the backend, I'll create a separarePR for that once this one is done.

josephdadams commented:
Great! I had started converting all the source types to their own classes but not in ts. I’ll push that to its own branch in case it is helpful. My thinking was to try to abstract it as much as possible so all the source specific logic was contained within that file.

hrueger commented:
Thanks @josephdadams I'll take a look at that.
Another question: Do you mind if I move the version (at the bottom of the settings page) and the license (at the welcome page) to a separate about page? This way, I could redesign the welcome page (for example with some big buttons to go to the tally / producer / settings page. Or do you prefer having the license directly visible on the welcome page?
We could also go with `For licenense information go to the about page`.

josephdadams commented:
It’s fine with me to move those to another page. I’m not particular about that at all. Thanks!

hrueger commented:
The basic functionality is now complete. If anybody wants to try it, just checkout the `master` branch of my fork (`git clone https://github.com/hrueger/TallyArbiter` and install the dependencies with `npm i`. Then, start the server as normal (`node index.js`).

To start the new UI you first need to install its dependencies in the `UI` directory (`cd UI && npm i`). Then, run the UI with `ng serve` from the `UI` folder. You might need to install the Angular CLI if the `ng` command is not recognized (`npm i -g @angular/cli`). Once ready, you can open a browser to `localhost:4200`.

josephdadams commented on index.js:
```diff
@@ -5249,10 +5249,12 @@ function StartTSLClientConnection(tslClientId) {
-						tsl_clients[i].connected = false;
-						UpdateSockets('tsl_clients');
+						if (tsl_clients[i]) {
+							logger(`TSL Client ${tslClientId} Connection Closed: ${tsl_clients[i]?.ip}:${tsl_clients[i]?.port}`, 'info-quiet');
```
I had to remove these question marks `tsl_clients[i]?.ip` to `tsl_clients[i].ip` for it to work

hrueger replied:
Oops, should be fixed. You're probably using Node 13 or lower as optional chaining arrived with Node v14. I had to add the if condition because the server would crash when removing a disconnected TSL Client. If you want, I can revert this change as it is beyond the scope of this PR.

josephdadams replied:
I’m ok with it as long as we make sure people know they need to upgrade node. Companion requires 12.21 so generally that’s what I use for other stuff also.

hrueger replied:
Sounds good, but with 97e9b0a we should be compatible again anyway. The question marks and the if condition were Kind of redundant, so I removed them again. 👍

hrueger commented:
All the functionality is now complete.
You can also now run TallyArbiter as a CLI. The tarball is generated with GitHub actions. It is (temporarily) available to download here: https://github.com/hrueger/TallyArbiter/suites/2833604095/artifacts/63171946
Just unzip it and install it globally with `npm i -g TallyArbiter-1.5.9.tgz` Then, run it by typing `tallyarbiter` in a terminal.

josephdadams commented:
Incredible work! I will start looking it over here in the next few days as I have time. Really excited about the improvements you've made here, and I think a lot of people will benefit from this. I'm not familiar with Angular at all, so a lot of it I will just take your word for it, as it's something I'll have to learn on my own time.

Do you have other steps on your list before you would want to see this committed?

josephdadams commented:
<img width="1173" alt="Screen Shot 2021-05-26 at 10 07 01 AM" src="https://user-images.githubusercontent.com/2559091/119674840-87aacc00-be0a-11eb-9a27-50ece71e6f09.png">
Just wanted to mention quickly a couple things that I saw on the MacOS version.
* the sys tray icon looks like this for some reason
* Data doesn't persist between sessions, i.e. if I add a source type, when I close out and re-open the electron app, that source type is not there. I'm sure this is because of the way TA stores data in the local file that electron doesn't have access to, so we probably just need to use a different store method for electron, and possibly implement an import/export feature for existing users

hrueger commented:
Thanks! Take your time.

> Do you have other steps on your list before you would want to see this committed?

Yes, there are definitely some.
- [ ] I think, we should get some more people to test this.
- [ ] If you're ok with that, I'd setup the GitHub Actions workflow to automatically create a GitHub Release once someone pushes a git tag. It could also publish the `tallyarbiter` package to npm. But for that I'd need an npm access token added to this repo's environment secrets.
- [ ] Fix the Tray Icon
- [ ] Fix the storage

Regarding the config: What do you think about saving the `config.json` file in the application data directory? Electron provides the [path](https://www.electronjs.org/docs/api/app#appgetpathname) but we should also be able to get this in pure JS. If we have a common storage file, people could switch from the CLI to a local dev version (from `git clone`) or to the electron version and vice versa.

hrueger commented:
### Download latest build artifacts
As I don't want to paste the new links to the installers and the tarball here every time I make a commit, please download the files like so until this PR is complete:
1. Go to https://github.com/hrueger/TallyArbiter/actions
2. Click on the topmost workflow run with the ✔ icon
3. At the bottom of that page you can download the artifacts.

josephdadams commented:
> Regarding the config: What do you think about saving the `config.json` file in the application data directory? Electron provides the [path](https://www.electronjs.org/docs/api/app#appgetpathname) but we should also be able to get this in pure JS. If we have a common storage file, people could switch from the CLI to a local dev version (from `git clone`) or to the electron version and vice versa.

Sure! Happy to save it wherever it works as long as that method works across all platforms (pi included).

josephdadams commented:
Also, feel free to contribute any changes to the wiki or readme that feel applicable for this PR.

hrueger commented:
I added that to the todo list 👍

josephdadams commented:
Also when we merge this I feel it is a significant change to the codebase that I’d like to version bump to 2.0

hrueger commented:
Sounds good 👍

The config is now saved like so:
- Windows: `C:\Users\YourUsername\AppData\Roaming\TallyArbiter\config.json`
- MacOS: `~/Library/Preferences/TallyArbiter/config.json`
- Linux: `~/.local/share/TallyArbiter/config.json`

If any of the folders in the path don't exist, they will be created recursively.

I just tried it with the CLI and the Electron Version and they both now read and write to the same shared config file (however, you should not run both versions at the same time but I don't think anybody does that).

I also added the missing chat to the tally page. However, it does not receive any messages. I need to do some debugging there.

josephdadams commented:
Great job!

hrueger commented:
Thanks. The chat on the Tally page should work now, too.

hrueger commented:
@josephdadams I updated the GitHub Action workflows to automatically create releases and publish the CLI to npm once a tag is pushed. You can do that by first changing the version in the `package.json` and `package-lock.json` files (either manually or with `npm version`) and then typing `git tag v2.0.0` followed by `git push --tags`.

In order for the npm publish to work, you will need to add a repository secret in the settings. It must be called `NPM_ACCESS_TOKEN` and the value must be an npm access token.

I also updated the Readme. As my English is probably not the best, feel free to correct any spelling errors, wrong sentences, etc.

Regarding the Wiki: I honestly don't think it's a good idea to have documentation both in the Readme and the wiki. And I'm (personally) not a fan of the GitHUb Wiki. Actually, I didn't even recognize that there was content in the wiki when I was first using TallyArbiter.

What do you think if I build a nice little docs site using [Docusaurus](https://docusaurus.io) which will be hosted on GitHub pages?

If not, somebody else has to edit the wiki, because I can't even clone it on Windows. There seem to be some special characters in the filenames.

josephdadams commented:
I’m great with docusaurus! The plan all along had been to greatly simplify the readme to just be install/build instructions as it relates to contributing to the project, with other documentation elsewhere for how to use the project.

josephdadams commented:
That looks great! I think it will make it a lot easier for people to follow, and with it being hosted on Github, perhaps others can contribute to the documentation as the project continues on.

hrueger commented:
I added information about the new UI to the README and also a little guide on how to contribute / setup the development environment. I believe that this PR is now complete, only the NPM Auth Token is missing (see above)

josephdadams commented:
Really great work, @hrueger! I'm excited for the direction you've set TA. It will be much easier for users now, and more manageable as well. I will continue reviewing and ping you as I see things. In the meantime, if you see things you'd like to improve upon, feel free to add.

josephdadams commented:
@hrueger Electron users might miss "important" data that is currently logged `info-quiet`. Ideas on how we could let them get access to that? It's fine with me if we just log all of that in the UI. (Change all `info-quiet` to `info`).

hrueger commented:
What about adding a select box with the options "error", "info", "verbose" to the logs? Then, users could choose what they want to see.

josephdadams commented:
Excellent!

hrueger commented:
I just fixed that on Windows TallyArbiter would just have a blank application icon. Strangely, this is due to the `description` field in the `package.json` being too long... (see [this comment](https://github.com/electron-userland/electron-builder/issues/2435#issuecomment-380596113)). So I had to shorten the description. I hope, you don't mind.

josephdadams commented:
I don't mind. Thanks!

josephdadams commented:
I noticed that socketio has been bumped higher than v2.3. If I recall, everything v3 and higher is not compatible with v2. So this would be a breaking change for all of the existing listener clients, and something to consider for an upgrade plan on that end of things.

hrueger commented:
I enabled the [compatibility mode](https://socket.io/docs/v4/server-initialization/#allowEIO3) which should allow v2 and v3 clients to connect to the v4 server.

josephdadams commented:
Didn't even know that was a thing. Thanks!

hrueger commented:
That would be great, thanks 👍

hrueger commented:
Hm. I downgraded `packet` to `0.0.6` but it's still not working. There's no problem if I run the electron version locally with `npm run desktop, but if I build the installer and run the installed version, it's not working. I'll do some more tests.

hrueger commented:
I can, however, connect to an OBS on another computer, so I was probably wrong and it is not a firewall issue.

hrueger commented:
That's the problem: https://github.com/nrkno/tv-automation-atem-connection/issues/106#issuecomment-860174110

hrueger commented:
Issue should be fixed now, for Electron the `3.0.0-nightly` versions of `atem-connection` have to be used.

I also added the so-called single instance lock for Electron which prevents users from opening TallyArbiter more than once and instead shows and focuses the already open window.

JTF4 commented:
While testing I saw that there is no way to manually define the TSL return address on a device. This was available in the old UI.

hrueger commented:
Interesting, I thought that I had added all option fields in all dialogs. I'll check that in a few days.

josephdadams commented:
@hrueger I've created the npm package for tallyarbiter as well as added the token to the repo settings. So unless I missed something, we should be good on that end when this gets merged. We (@JTF4) will hopefully be using this in a production test tomorrow and will report back any breaks.

hrueger commented:
Sounds great 👍

josephdadams commented:
Test went great. I’ll be looking to merge this week. Anything major left before I do so?

hrueger commented:
Great 👍 I don't think that there's anything left. Once it's merged you just need to enable GitHub Pages for the `gh-pages` branch if you haven't done that yet.

josephdadams commented:
@hrueger will the gh-pages branch automatically come over with this PR, or do we need a separate PR for that?

hrueger commented:
It should be created automatically.

josephdadams commented:
@hrueger I've merged the PR. Fantastic job. Thanks so much!

It looks like the actions are running and the builds are being created. I've created a new tag for release which fires off the actions, but not seeing the build files showing up with the release. What am I missing?

hrueger commented:
@josephdadams Great news 😀

> What am I missing?

I always create the tag using the command line with `git tag v2.0.0` and then push it with `git push --tags`. This triggers the action which automatically creates a draft release. The artifacts are then added to the draft release. Once that has finished, I add my changelog texts to the Release and publish it.

The upload process is handled entirely by `electron-builder`, so I don't know if it can add artifacts to already published releases...
I assume that you created the release which auto-created the tag.

You could try creating the tag `v2.0.0` using the command line. I generally use the same tag version as the npm version, so that would be `v2.0.0` instead of `v2.0`.

josephdadams commented:
I created the release through the web interface.

I took at look at the build.yml, it looks like all the "upload" parts are commented out. Is that correct?

josephdadams commented:
Ok, I clearly just did something wrong, because gh-actions had a draft release of 2.0.0 for me. I released that and it had the binaries attached as expected. Sorry for my naivety, I've not done anything with actions up to the point. Thanks!

hrueger commented:
No problem, glad it works now 👍

hrueger commented:
>I took at look at the build.yml, it looks like all the "upload" parts are commented out. Is that correct?

This upload adds build artifacts directly to the workflow run. I found that quite useful for testing but I don't think this is needed in production.

MatteoGheza commented:
@hrueger since GH releases currently work, can you implement Electron autoUpdater?

hrueger commented:
Sure 👍

mattv8 commented:
Had a chance to look at the Angular frontend, looks fantastic. Amazing work @hrueger and @josephdadams!! 👐👐

mattv8 commented:
Just as an FYI: after some more testing trying to get the Electron app running on a Windows environment (building the TA and Desktop UI from source) I get the error generated by Electron:
`elfy@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter$ npm run desktop

> tallyarbiter@2.0.0 desktop
> electron .

[16432:0708/133256.634715:FATAL:setuid_sandbox_host.cc(158)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter/node_modules/electron/dist/chrome-sandbox is owned by root and has mode 4755.`

Windows Subsystem for Linux has no root user to chmod the directory chrome-sandbox to, so unfortunately the build stops there. Are there plans for publishing pre-built releases for the Electron app in the future for us Windows laypeople?

MatteoGheza commented:
Have you tried removing _node_modules_ and re-running `npm install` in UI?

mattv8 commented:
Just tried it, here's what I see:
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter# rm -rf node_modules
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter# cd UI
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter/UI# npm install
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular-devkit/architect@0.1200.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular-devkit/build-angular@12.0.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular-devkit/build-optimizer@0.1200.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular-devkit/build-webpack@0.1200.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular-devkit/core@12.0.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular-devkit/schematics@12.0.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/animations@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/cli@12.0.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/common@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/compiler@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/compiler-cli@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/core@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/forms@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/platform-browser@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/platform-browser-dynamic@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/router@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@angular/service-worker@12.0.1',
>npm WARN EBADENGINE   required: { node: '^12.14.1 || ^14.0.0' },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@ngtools/webpack@12.0.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>npm WARN EBADENGINE Unsupported engine {
>npm WARN EBADENGINE   package: '@schematics/angular@12.0.1',
>npm WARN EBADENGINE   required: {
>npm WARN EBADENGINE     node: '^12.14.1 || ^14.0.0',
>npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6',
>npm WARN EBADENGINE     yarn: '>= 1.13.0'
>npm WARN EBADENGINE   },
>npm WARN EBADENGINE   current: { node: 'v15.14.0', npm: '7.7.6' }
>npm WARN EBADENGINE }
>
>up to date, audited 1342 packages in 5s
>
>11 vulnerabilities (6 moderate, 5 high)
>
>To address issues that do not require attention, run:
>  npm audit fix
>
>Some issues need review, and may require choosing
>a different dependency.
>
>Run `npm audit` for details.
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter/UI# npm run desktop
>npm ERR! Missing script: "desktop"
>npm ERR!
>npm ERR! To see a list of scripts, run:
>npm ERR!   npm run
>
>npm ERR! A complete log of this run can be found in:
>npm ERR!     /root/.npm/_logs/2021-07-08T19_48_49_872Z-debug.log
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter/UI# cd ../
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter# npm run desktop
>
>> tallyarbiter@2.0.0 desktop
>> electron .
>
>sh: 1: electron: not found
>root@Matt-Surface:/mnt/c/Users/mattv/Downloads/Git Repositories/TallyArbiter#

MatteoGheza commented:
Try removing _node_modules_ in the root directory of the project and in UI and then run `npm install` in the root directory.

hrueger commented:
@mattv8 Just that I understand what you are trying to accomplish: You are inside Ubuntu on WSL2 and are trying to run the Electron version from inside?

And there is a Windows Installer available for download on the Releases page. Or are you looking for a portable version?

mattv8 commented:
@hrueger that's exactly right, Ubuntu 18.0.4 WSL2. I'm just trying to understand how to build the project for myself.

Also, I'm a goober and didn't notice the releases page here: https://github.com/josephdadams/TallyArbiter/releases Typically on a repo you see the releases from the project root (https://github.com/josephdadams/TallyArbiter) on the right column so I had just missed that Joseph created a Releases page.

@MatteoGheza Tried as you suggested, and saw the same error as before:
>[17352:0708/141908.110016:FATAL:setuid_sandbox_host.cc(158)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /usr/lib/node_modules/electron/dist/chrome-sandbox is owned by root and has mode 4755.
/usr/lib/node_modules/electron/dist/electron exited with signal SIGTRAP

hrueger commented:
I have not yet managed to run gui apps and therefore electron apps inside WSL2. I've read that support for those is coming with one of the next Windows Updates.
So I don't know if you can run it inside Ubuntu in WSL2. 

But you could try building and running on Windows?

mattv8 commented:
I will have to figure out how to build and run on Windows, as I'm unfamiliar with NodeJS on anything other than Linux & derivatives. I'll give it a try and report back.

hrueger commented:
Did you try `--no-sandbox`?

By the way: running `electron .` is not actually building the app, but instead just running the electron binary with the path to the package directory as the first argument. The only building we have is the build of the Angular Frontend and the build of the installers.

hrueger commented:
@josephdadams can you enable the "Releases" tab in the repo settings somewhere? I wonder why it does not appear.

hrueger commented:
> I will have to figure out how to build and run on Windows, as I'm unfamiliar with NodeJS on anything other than Linux & derivatives. I'll give it a try and report back.

Good luck, it should actually not be much different. You just need to make sure that you run `npm i` from the Windows env again (after you have deleted the `node_modules` folder as then the electron installer script will download the correct binaries. Otherwise you still have Linux binaries under Windows and I don't think that this works well 😉

josephdadams commented:
@hrueger I added you to the repo to see if you can find it because I could not :)

hrueger commented:
Thanks. I don't have permissions to do that, but I think this is the button you need to click on:
![Screenshot_20210708-223728_Firefox.jpg](https://user-images.githubusercontent.com/18656830/124987549-4890ad00-e03d-11eb-848e-3df13f4c682f.jpg)

In that modal you should be able to avtivate the releases tab.

Sorry for the bad quality, but I'm just on my phone.

josephdadams commented:
I figured it out with your help :)
